DeviceAdapters/Zaber/ConnectionManager.cpp

```diff
@@ -0,0 +1,39 @@
+#include "ConnectionManager.h"
+
+std::shared_ptr<zml::Connection> ConnectionManager::getConnection(std::string port, MM::Core* core)
+{	
+	std::lock_guard<std::mutex> lockGuard(lock_);
+	if (connections_.count(port)) {
+		if (auto connectionPtr = connections_.at(port).lock()) {
+			return connectionPtr;
+		}
+	}
+
+	auto connection = std::make_shared<zml::Connection>(zml::Connection::openSerialPort(port));
+	auto id = connection->getInterfaceId();
+	connection->getDisconnected().subscribe([=](std::shared_ptr<zmlbase::MotionLibException>) {
+		removeConnection(port, id);
+	});
+	connections_[port] = connection;
+
+	return connection;
+}
+
+bool ConnectionManager::removeConnection(std::string port, int interfaceId)
+{
+	std::lock_guard<std::mutex> lockGuard(lock_);
+	if (!connections_.count(port)) {
+		return false;
+	} 
+
+	if (interfaceId != -1) {
+		if (auto connection = connections_.at(port).lock()) {
+			if (connection->getInterfaceId() != interfaceId) {
```

Comment:
Why do you need this ID check?

Reply:
To prevent the race condition when I remove a new connection that was just added as replacement.

Reply:
It's unlikely but not impossible.